hfsutils/doc/man/hformat.1
.TH HFORMAT 1 08-Nov-1997 HFSUTILS
.SH NAME
hformat \- create a new HFS filesystem and make it current
.SH SYNOPSIS
hformat
[-f] [-l
.IR label ]
.I destination-path
.RI [ partition-no ]
.SH DESCRIPTION
.B hformat
is used to write a new HFS filesystem to a volume. A UNIX pathname to the
volume's destination must be specified. The destination may be either a block
device or a regular file, but it must already exist and be writable.
.PP
An optional label can be specified to name the volume. The name must be
between 1-27 characters and cannot contain a colon (:). By default, the volume
will be named
.BR Untitled .
.PP
If the destination medium is partitioned, one partition must be selected to
receive the filesystem. If there is only one HFS partition on the medium, it
will be selected by default. Otherwise, the desired partition number must be
specified (as the ordinal
.IR n th
HFS partition) on the command-line. The size of the partition determines the
size of the resulting volume.
.PP
Partition number
.B 0
can be specified to format the entire medium as a single filesystem without a
partition map, erasing any existing partition information. Since this will
destroy all the partitions, the
.B -f
option must be specified to force this operation if the medium currently
contains a partition map.
.PP
If the medium is not partitioned (or if partition 0 is specified), the size or
capacity of the medium determines the size of the resulting volume.
.PP
The new volume will be empty and will become "current" so subsequent commands
will refer to it. The current working directory for the volume is set to the
root of the volume.
.SH EXAMPLES
.SP
.TP
% hformat /dev/fd0
If a floppy disk is available as
.BR /dev/fd0 ,
this formats the disk as an HFS volume named
.BR Untitled .
(N.B. The floppy must already have received a low-level format by other
means.)
.TP
% dd if=/dev/zero of=disk.hfs bs=1k count=800
.PD 0
.TP
% hformat -l "Test Disk" disk.hfs
.PD 1
This sequence creates an 800K HFS volume image in the file
.B
disk.hfs
in the current directory, and names it
.BR "Test Disk" .
.TP
% hformat -l "Loma Prieta" /dev/sd2 1
If a SCSI disk is available as
.BR /dev/sd2 ,
this initializes the first HFS partition on the disk (which must already
exist) with a new filesystem, naming the resulting volume
.BR "Loma Prieta" .
.TP
% hformat -f /dev/sd2 0
This causes the medium accessible as
.B /dev/sd2
to be reformatted as a single HFS volume, ignoring and erasing any existing
partition information on the medium. The
.B -f
option must be specified if the medium is currently partitioned; otherwise the
command will fail.
.SH NOTES
This command does not create or alter partition maps, although it can erase
them (as described above). Any partition number specified on the command line
must already exist.
.PP
The smallest volume size which can be formatted with
.B hformat
is 800K.
.SH SEE ALSO
hfsutils(1), hmount(1)
.SH FILES
$HOME/.hcwd
.SH AUTHOR
Robert Leslie <rob@mars.org>


gcc/NEWS
This file contains information about GCC releases which has been generated
automatically from the online release notes.  It covers releases of GCC
(and the former EGCS project) since EGCS 1.0, on the line of development
that led to GCC 3. For information on GCC 2.8.1 and older releases of GCC 2,
see ONEWS.

======================================================================
http://gcc.gnu.org/gcc-4.7/index.html

                             GCC 4.7 Release Series

   March 22, 2012

   The [1]GNU project and the GCC developers are pleased to announce the
   release of GCC 4.7.0.

   This release is a major release, containing new features (as well as
   many other improvements) relative to GCC 4.6.x.

Release History

   GCC 4.7.0
          March 22, 2012 ([2]changes)

References and Acknowledgements

   GCC used to stand for the GNU C Compiler, but since the compiler
   supports several other languages aside from C, it now stands for the
   GNU Compiler Collection.

   A list of [3]successful builds is updated as new information becomes
   available.

   The GCC developers would like to thank the numerous people that have
   contributed new features, improvements, bug fixes, and other changes as
   well as test results to GCC. This [4]amazing group of volunteers is
   what makes GCC successful.

   For additional information about GCC please refer to the [5]GCC project
   web site or contact the [6]GCC development mailing list.

   To obtain GCC please use [7]our mirror sites or [8]our SVN server.


    For questions related to the use of GCC, please consult these web
    pages and the [9]GCC manuals. If that fails, the
    [10]gcc-help@gcc.gnu.org mailing list might help. Comments on these
    web pages and the development of GCC are welcome on our developer
    list at [11]gcc@gcc.gnu.org. All of [12]our lists have public
    archives.

   Copyright (C) [13]Free Software Foundation, Inc. Verbatim copying and
   distribution of this entire article is permitted in any medium,
   provided this notice is preserved.

   These pages are [14]maintained by the GCC team. Last modified
   2012-03-22[15].

References

   1. http://www.gnu.org/
   2. http://gcc.gnu.org/gcc-4.7/changes.html
   3. http://gcc.gnu.org/gcc-4.7/buildstat.html
   4. http://gcc.gnu.org/onlinedocs/gcc/Contributors.html
   5. http://gcc.gnu.org/index.html
   6. mailto:gcc@gcc.gnu.org
   7. http://gcc.gnu.org/mirrors.html
   8. http://gcc.gnu.org/svn.html
   9. http://gcc.gnu.org/onlinedocs/
  10. mailto:gcc-help@gcc.gnu.org
  11. mailto:gcc@gcc.gnu.org
  12. http://gcc.gnu.org/lists.html
  13. http://www.fsf.org/
  14. http://gcc.gnu.org/about.html
  15. http://validator.w3.org/check/referer
======================================================================
http://gcc.gnu.org/gcc-4.7/changes.html

                             GCC 4.7 Release Series
                        Changes, New Features, and Fixes

Caveats

     * Support for a number of older systems and recently unmaintained or
       untested target ports of GCC has been declared obsolete in GCC 4.7.
       Unless there is activity to revive them, the next release of GCC
       will have their sources permanently removed.
       The following ports for individual systems on particular
       architectures have been obsoleted:
          + IRIX 6.5 (mips-sgi-irix6.5)
          + MIPS OpenBSD (mips*-*-openbsd*)
          + Solaris 8 (*-*-solaris2.8). Details can be found in the
            [1]announcement.
          + Tru64 UNIX V5.1 (alpha*-dec-osf5.1*)
     * Support on ARM for the legacy floating-point accelerator (FPA) and
       the mixed-endian floating-point format that it used has been
       obsoleted. The ports that still use this format have been obsoleted
       as well. Many legacy ARM ports already provide an alternative that
       uses the VFP floating-point format. The obsolete ports will be
       deleted in the next release.
       The obsolete ports with alternatives are:
          + arm*-*-rtems (use arm*-*-rtemseabi)
          + arm*-*-linux-gnu (use arm*-*-linux-gnueabi)
          + arm*-*-elf (use arm*-*-eabi)
          + arm*-*-uclinux* (use arm*-*-uclinux*eabi)
       Note, however, that these alternatives are not binary compatible
       with their legacy counterparts (although some can support running
       legacy applications).
       The obsolete ports that currently lack a modern alternative are:
          + arm*-*-ecos-elf
          + arm*-*-freebsd
          + arm*-wince-pe*
       New ports that support more recent versions of the architecture are
       welcome.
     * Support for the Maverick co-processor on ARM has been obsoleted.
       Code to support it will be deleted in the next release.
     * Support has been removed for Unix International threads on Solaris
       2, so the --enable-threads=solaris configure option and the
       -threads compiler option don't work any longer.
     * Support has been removed for the Solaris BSD Compatibility Package,
       which lives in /usr/ucbinclude and /usr/ucblib. It has been removed
       from Solaris 11, and was only intended as a migration aid from
       SunOS 4 to SunOS 5. The -compat-bsd compiler option is not
       recognized any longer.
     * The AVR port's libgcc has been improved and its multilib structure
       has been enhanced. As a result, all objects contributing to an
       application must either be compiled with GCC versions up to 4.6.x
       or with GCC versions 4.7.0 or later.
     * The ARM port's -mwords-little-endian option has been deprecated. It
       will be removed in a future release.
     * Support has been removed for the NetWare x86 configuration
       obsoleted in GCC 4.6.
     * It is no longer possible to use the "l" constraint in MIPS16 asm
       statements.
     * More information on porting to GCC 4.7 from previous versions of
       GCC can be found in the [2]porting guide for this release.

General Optimizer Improvements


hfsutils/hformat.c
/*
 * hfsutils - tools for reading and writing Macintosh HFS volumes
 * Copyright (C) 1996-1998 Robert Leslie
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * $Id: hformat.c,v 1.9 1998/11/02 22:08:31 rob Exp $
 */

# ifdef HAVE_CONFIG_H
#  include "config.h"
# endif

# ifdef HAVE_UNISTD_H
#  include <unistd.h>
# endif

# include <stdio.h>
# include <stdlib.h>
# include <string.h>

# include "hfs.h"
# include "hcwd.h"
# include "hfsutil.h"
# include "suid.h"
# include "hformat.h"

# define O_FORCE	0x01

extern char *optarg;
extern int optind;

/*
 * NAME:	usage()
 * DESCRIPTION:	display usage message
 */
static
void usage(void)
{
  fprintf(stderr, "Usage: %s [-f] [-l label] path [partition-no]\n", argv0);
}

/*
 * NAME:	do_format()
 * DESCRIPTION:	call hfs_format() with necessary privileges
 */
static
hfsvol *do_format(const char *path, int partno, int mode, const char *vname)
{
  hfsvol *vol = 0;

  suid_enable();

  if (hfs_format(path, partno, mode, vname, 0, 0) != -1)
    vol = hfs_mount(path, partno, HFS_MODE_ANY);

  suid_disable();

  return vol;
}

/*
 * NAME:	hformat->main()
 * DESCRIPTION:	implement hformat command
 */
int hformat_main(int argc, char *argv[])
{
  const char *vname;
  char *path = 0;
  hfsvol *vol;
  hfsvolent ent;
  int nparts, partno, options = 0, result = 0;

  vname = "Untitled";

  while (1)
    {
      int opt;

      opt = getopt(argc, argv, "fl:");
      if (opt == EOF)
	break;

      switch (opt)
	{
	case '?':
	  usage();
	  goto fail;

	case 'f':
	  options |= O_FORCE;
	  break;

	case 'l':
	  vname = optarg;
	  break;
	}
    }

  if (argc - optind < 1 || argc - optind > 2)
    {
      usage();
      goto fail;
    }

  path = hfsutil_abspath(argv[optind]);
  if (path == 0)
    {
      fprintf(stderr, "%s: not enough memory\n", argv0);
      goto fail;
    }

  suid_enable();
  nparts = hfs_nparts(path);
  suid_disable();

  if (argc - optind == 2)
    {
      partno = atoi(argv[optind + 1]);

      if (nparts != -1 && partno == 0)
	{
	  if (options & O_FORCE)
	    {
	      fprintf(stderr,
		      "%s: warning: erasing partition information\n", argv0);
	    }
	  else
	    {
	      fprintf(stderr, "%s: medium is partitioned; "
		      "select partition > 0 or use -f\n", argv0);
	      goto fail;
	    }
	}
    }
  else
    {
      if (nparts > 1)
	{
	  fprintf(stderr,
		  "%s: must specify partition number (%d available)\n",
		  argv0, nparts);
	  goto fail;
	}
      else if (nparts == -1)
	partno = 0;
      else
	partno = 1;
    }

  vol = do_format(path, partno, 0, vname);
  if (vol == 0)
    {
      hfsutil_perror(path);
      goto fail;
    }

  hfs_vstat(vol, &ent);
  hfsutil_pinfo(&ent);

  if (hcwd_mounted(ent.name, ent.crdate, path, partno) == -1)
    {
      perror("Failed to record mount");
      result = 1;
    }

  hfsutil_unmount(vol, &result);

  free(path);

  return result;

fail:
  if (path)
    free(path);

  return 1;
}
